src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -1345,13 +1349,45 @@ private void computeChangeLog(GitClient git, Revision revToBuild, TaskListener l
             }
 
             if (!exclusion) {
-                // this is the first time we are building this branch, so there's no base line to compare against.
-                // if we force the changelog, it'll contain all the changes in the repo, which is not what we want.
-                listener.getLogger().println("First time build. Skipping changelog.");
-            } else {
-                changelog.to(out).max(MAX_CHANGELOG).execute();
-                executed = true;
+                listener.getLogger().println("First build of this branch, computing origin.");
+                // Ok, so here we are naked ... All we have is a detached head, on a SHA1 checkout, with no informations on remote heads/refs
+                // So we first list all known SHA1's from jobs having at least one remote url in common
+                // Then we go down current rev's ancestors until we found one already built, this is our stop
+
+                Set<String> ourUrls = userRemoteConfigs.stream().map(c -> c.getUrl()).collect(Collectors.toSet());
+
+                Instant start = Instant.now();
+                List<ObjectId> knownSHA1 = (List<ObjectId>) Jenkins.get().getItems().stream()
+                    .flatMap(i -> i.getAllJobs().stream())
+                    .flatMap(j -> j.getBuilds().stream())
```

Comment:
This is unacceptable in general. Basically since 1.484 onwards, you may only call `getBuilds()` (or do related things like loop over build numbers) if you have a very strict bound before giving up (this list is lazy) _or_ are incrementally producing output for `ProgressiveRendering` or another system with built-in paging / cancellation semantics. Otherwise you will wind up loading and parsing an endless series of `build.xml` files into heap, until memory runs out at which point `SoftReference`s will be cleared, that work will be discarded, and you will start over.